README.md
# simple_carousel

A simple carousel package.

[PubDev](https://pub.dev/packages/simple_carousel)

## Instalation

Add `simple_carousel: 0.0.1` in your `pubspec.yaml`.

## Motivation

This simple widget that will improve with time. If your collaborate please access [github project](https://github.com/FabioFiuza/simple-carousel) project and send a pull request.

## How to use

The parent widget should have a size.

```
class FruitCarouselPage extends StatelessWidget {
  FruitCarouselPage({Key key}) : super(key: key);

  final fruits = <String>['orange', 'apple', 'banana', 'strawberry'];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Fruit Carousel'),
      ),
      body: Container(
        margin: EdgeInsets.all(16),
        height: 150,
        child: SimpleCarousel(
          colorIconCircleBar: Colors.orange[100],
          colorIconCircleBarActive: Colors.orange,
          numberPages: fruits.length,
          children: fruits
              .map(
                (fruit) => FruitCarouselItem(
                  fruitName: fruit,
                ),
              )
              .toList(),
        ),
      ),
    );
  }
}

class FruitCarouselItem extends StatelessWidget {
  const FruitCarouselItem({Key key, this.fruitName}) : super(key: key);

  final String fruitName;

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(color: Colors.grey[300]),
      child: Center(
        child: Text(
          fruitName.toUpperCase(),
          style: Theme.of(context).textTheme.subtitle1.copyWith(
                fontWeight: FontWeight.bold,
              ),
        ),
      ),
    );
  }
}
```

[![example](https://i.postimg.cc/nz5XZzG7/Captura-de-Tela-2020-10-18-a-s-14-36-32.png)](https://postimg.cc/476fQXMN)


CHANGELOG.md
## [0.0.1] - 2020-10-18

- Added simple creation of the carousel widget using Flutter pageView.


test/simple_carousel_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_carousel/simple_carousel.dart';

void main() {
  Widget simpleCarousel;

  setUp(() {
    simpleCarousel = MaterialApp(
      home: Scaffold(
        body: Container(
          height: 200,
          child: SimpleCarousel(
            numberPages: 2,
            children: [
              Center(
                key: ValueKey('item1'),
                child: Text('Teste1'),
              ),
              Center(
                child: Text('Teste2'),
              ),
            ],
          ),
        ),
      ),
    );
  });

  testWidgets('Simple carousel shoud be open successfully',
      (WidgetTester tester) async {
    await tester.pumpWidget(simpleCarousel);
    expect(find.byKey(ValueKey('circleBar0')), findsOneWidget);
    expect(find.byKey(ValueKey('circleBar1')), findsOneWidget);
    expect(find.text('Teste1'), findsOneWidget);
  });

  testWidgets('If the user makes a scroll the next item should is open',
      (WidgetTester tester) async {
    await tester.pumpWidget(simpleCarousel);
    await tester.drag(
        find.byKey(ValueKey('PageViewKey')), const Offset(-200.0, 0.0));
    await tester.pump(Duration(milliseconds: 100));

    expect(find.text('Teste2'), findsOneWidget);
  });
}


lib/simple_carousel.dart
library simple_carousel;

import 'package:flutter/material.dart';

/// Import that the parent widget has a size.
class SimpleCarousel extends StatefulWidget {
  const SimpleCarousel({
    Key key,
    @required this.numberPages,
    @required this.children,
    this.colorIconCircleBarActive,
    this.colorIconCircleBar,
  })  : assert(numberPages != null),
        assert(children != null),
        super(key: key);

  final int numberPages;

  final List<Widget> children;

  /// Default color is [Colors.grey[700]].
  final Color colorIconCircleBarActive;

  /// Default color is [Colors.grey].
  final Color colorIconCircleBar;

  @override
  _SimpleCarousel createState() => _SimpleCarousel();
}

class _SimpleCarousel extends State<SimpleCarousel> {
  PageController _controller;
  int _indexPage = 0;

  @override
  void initState() {
    _controller = PageController(initialPage: 0);
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Flexible(
          child: PageView(
            key: ValueKey('PageViewKey'),
            children: widget.children,
            physics: ClampingScrollPhysics(),
            onPageChanged: (index) {
              setState(() {
                _indexPage = index;
              });
            },
          ),
        ),
        SizedBox(
          height: 8,
        ),
        _buildPageIndicator(
          colorIconCircleBar: widget.colorIconCircleBar ?? Colors.grey,
          colorIconCircleBarActive:
              widget.colorIconCircleBarActive ?? Colors.grey[700],
          itemCount: widget.numberPages,
          currentPageValue: _indexPage,
        ),
      ],
    );
  }

  Widget _buildPageIndicator({
    final int currentPageValue,
    final int itemCount,
    final Color colorIconCircleBarActive,
    final Color colorIconCircleBar,
  }) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        for (int i = 0; i < itemCount; i++)
          if (i == currentPageValue) ...[
            _circleBar(
              key: ValueKey('circleBar$i'),
              isActive: true,
              colorIconCircleBar: colorIconCircleBar,
              colorIconCircleBarActive: colorIconCircleBarActive,
            )
          ] else
            _circleBar(
              key: ValueKey('circleBar$i'),
              isActive: false,
              colorIconCircleBar: colorIconCircleBar,
              colorIconCircleBarActive: colorIconCircleBarActive,
            ),
      ],
    );
  }

  Widget _circleBar({
    Key key,
    bool isActive,
    Color colorIconCircleBarActive,
    Color colorIconCircleBar,
  }) {
    return AnimatedContainer(
      key: key,
      duration: Duration(milliseconds: 150),
      margin: EdgeInsets.symmetric(horizontal: 8),
      height: isActive ? 12 : 8,
      width: isActive ? 12 : 8,
      decoration: BoxDecoration(
        color: isActive ? colorIconCircleBarActive : colorIconCircleBar,
        borderRadius: BorderRadius.all(Radius.circular(12)),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_carousel/simple_carousel.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Fruit carousel',
      theme: ThemeData(
        primarySwatch: Colors.orange,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: FruitCarouselPage(),
    );
  }
}

class FruitCarouselPage extends StatelessWidget {
  FruitCarouselPage({Key key}) : super(key: key);

  final fruits = <String>['orange', 'apple', 'banana', 'strawberry'];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Fruit Carousel'),
      ),
      body: Container(
        margin: EdgeInsets.all(16),
        height: 150,
        child: SimpleCarousel(
          colorIconCircleBar: Colors.orange[100],
          colorIconCircleBarActive: Colors.orange,
          numberPages: fruits.length,
          children: fruits
              .map(
                (fruit) => FruitCarouselItem(
                  fruitName: fruit,
                ),
              )
              .toList(),
        ),
      ),
    );
  }
}

class FruitCarouselItem extends StatelessWidget {
  const FruitCarouselItem({Key key, this.fruitName}) : super(key: key);

  final String fruitName;

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(color: Colors.grey[300]),
      child: Center(
        child: Text(
          fruitName.toUpperCase(),
          style: Theme.of(context).textTheme.subtitle1.copyWith(
                fontWeight: FontWeight.bold,
              ),
        ),
      ),
    );
  }
}
